plugins/net_plugin/net_plugin.cpp

```diff
@@ -2593,42 +2625,96 @@ namespace eosio {
       index.erase(p.first, p.second);
    }
 
-   bool dispatch_manager::add_peer_txn( const transaction_id_type& id, const time_point_sec& trx_expires,
-                                        uint32_t connection_id, const time_point_sec& now ) {
+   dispatch_manager::add_peer_txn_info dispatch_manager::add_peer_txn( const transaction_id_type& id, const time_point_sec& trx_expires, connection& c )
+   {
       fc::lock_guard g( local_txns_mtx );
-      auto tptr = local_txns.get<by_id>().find( std::make_tuple( std::ref( id ), connection_id ) );
-      bool added = (tptr == local_txns.end());
-      if( added ) {
+
+      auto& id_idx = local_txns.get<by_id>();
+      bool already_have_trx = false;
+      if (auto tptr = id_idx.find( id ); tptr != id_idx.end()) {
+         if (tptr->connection_ids.insert(c.connection_id).second)
+            ++c.trx_entries_size;
+         already_have_trx = tptr->have_trx;
```

Comment:
I feel that maybe we should update `tptr->expires` if `tptr->have_trx == false`.

Reply:
Each node has a configurable amount of time it will keep trxs entries around. If we update it then we are not really honoring that limit if they keep being sent to us.

Reply:
> Each node has a configurable amount of time it will keep trxs entries around. If we update it then we are not really honoring that limit if they keep being sent to us.

We store this expiration limit per transaction, not per connection. How can we honor the limit that each nodes sends?

Reply:
Each node does not send us an expiration. Our node has an expiration for how long it will remember a trx. I don't think another coming into the node should reset the expiration.

Reply:
Why have the expiration in the notice then? It is somewhat irrelevant imo. I think it would make sense to have the expiration only when we have the transaction (so maybe make it a `std::optional`?

Reply:
Often the expiration of the actual trx is going to be less than the configured node expiration. So it is sent along so that we can when appropriate remove it sooner.

Reply:
Yes, but why do we need it in the notice? I understand we need it in the transaction message.

Reply:
The extra few bytes for the expiration in the notice I doubt would make little difference but it does allow the receiver of the notice to potentially remove it sooner. Note there can/will be many notices sent where a trx is never sent. If the trx fails to subjectively execute on the node then it will never be sent to its peers. So imagine a node that is being spammed with trxs that fail. That node is going to spam all its peers with trx notice messages.

Reply:
> The extra few bytes for the expiration in the notice I doubt would make little difference

yes, I agree

> but it does allow the receiver of the notice to potentially remove it sooner

Sooner is a doubtful benefit.  If a node is spammed with transactions that fail, why would the spammer provide a short expiration time. 
I think maybe we should have a member in `node_transaction_state` which would be: `timestamp first_notice_received;`, and in `expire_txns`, we would remove all transactions where `have_trxn == false` and for which `now() - first_notice_received > max_trx_lifetime `

Reply:
In at least one common case of spam trxs they use a small (or default) expiration. If you are spamming a trx in as to hit some defi condition or win some game. You send many trx that fail. You are not trying to bring down the node, you are just trying to hit a condition on-chain.

Reply:
Actually since there is a potential large delay in the notice and actually receiving the trx, I can see updating the expiration when actually receiving the trx. I'll make that change.

Reply:
https://github.com/AntelopeIO/spring/pull/1499/commits/446fc5ce49c6eb0d22bad78795c5bdc8de22d11e